src/Build/Evaluation/Expander.cs

```diff
@@ -1108,12 +1108,18 @@ _metadata is IItemTypeDefinition itemMetadata &&
 
                         if (IsTruncationEnabled(_options) && metadataValue.Length > CharacterLimitPerExpansion)
                         {
-                            metadataValue =
-#if NET
-                                $"{metadataValue.AsSpan(0, CharacterLimitPerExpansion - 3)}...";
-#else
-                                $"{metadataValue.Substring(0, CharacterLimitPerExpansion - 3)}...";
-#endif
+                            // PERF: We need the formatted, truncated string. Using something like a StringBuilder avoids
+                            // needing to use an unsafe block, but this is more efficient.
```

Comment:
Is it really so much more efficient than

```c#
sb.Append(metadataValue, 0, CharacterLimitPerExpansion - 3);
sb.Append("...");
metadataValue = sb.ToString();
```

with a pooled StringBuilder? Using `unsafe` to hack at an immutable string feels very bad to me.

Reply:
It's roughly 2x the CPU cost since you have to copy the bytes to the internal `StringBuilder` and then to the final string. Does that matter? Eh, probably not. We've used this pattern in a handful of places similar to this, so I decided to pitch this idea first. It's only safe since we create a brand new string that we're the sole owner of while we do the unsafe mutation.

Reply:
`DefaultInterpolatedStringHandler` and `String.Concat()` should be fairly similar, but it looks like the interpolation may sometimes allocate twice (if its initial guessed allocation is too short): https://github.com/dotnet/runtime/blob/1d1bf92fcf43aa6981804dc53c5174445069c9e4/src/libraries/System.Private.CoreLib/src/System/Runtime/CompilerServices/DefaultInterpolatedStringHandler.cs#L135-L145

So I agree with Eric: `Concat(span, span)` is ~optimal here.